File: apps/mobile/shared/Sources/FlagshipCore/BoxTrustStatus.swift
import Foundation
import CryptoKit
import FlagshipAPI

/// Swift mirror of `packages/protocol/src/boxTrustStatus.ts` — canonical bytes
/// + Ed25519 verify for the box's STK-signed PER-BOX relay-trust verdict.
///
/// Each box independently verifies the ServiceBlessing it is handed by the
/// `.services` hub; that verdict is genuinely per-box. The daemon signs this
/// tuple with its STK; `.com` relays it VERBATIM on `/pods`
/// (`trustStatus: { report, signatureHex }`). A phone that derives the box
/// STK locally (`ServerKeys.deriveStkPub` — NOT `.com`'s echo) re-verifies
/// it end-to-end, so a rogue `.com` can DROP a report but cannot FORGE one:
/// the per-server warning a client renders is the box's own word.
///
/// SIBLING of `DaemonStatus` — do NOT fold together. Canonical bytes MUST
/// match the TS implementation byte-for-byte (pinned cross-platform vector in
/// packages/protocol/tests/boxTrustStatus.test.ts):
///
///   flagship/box-trust-status/v1|<serverDomain>|<relayVerdict>|
///   <lockedDown "1"|"0">|<failingCertHash or "">|
///   <coveringExceptionCertHash or "">|<nonce>|<issuedAt>

/// The box's verdict on the relay-class (`.services` hub) blessing it holds.
public enum RelayVerdict: String, Codable, Sendable {
    case trusted
    case untrusted
    case unknown
}

public struct BoxTrustStatusReport: Codable, Equatable, Sendable {
    public let serverDomain: String
    public let relayVerdict: RelayVerdict
    public let lockedDown: Bool
    /// relay-class cert-hash of the offending hub key, when untrusted.
    public let failingCertHash: String?
    /// relay-class cert-hash of the owner TrustException that lifted the
    /// failing verdict, when an override is in force.
    public let coveringExceptionCertHash: String?
    public let nonce: String
    public let issuedAt: Int64

    public init(
        serverDomain: String,
        relayVerdict: RelayVerdict,
        lockedDown: Bool,
        failingCertHash: String?,
        coveringExceptionCertHash: String?,
        nonce: String,
        issuedAt: Int64
    ) {
        self.serverDomain = serverDomain
        self.relayVerdict = relayVerdict
        self.lockedDown = lockedDown
        self.failingCertHash = failingCertHash
        self.coveringExceptionCertHash = coveringExceptionCertHash
        self.nonce = nonce
        self.issuedAt = issuedAt
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.serverDomain = try c.decode(String.self, forKey: .serverDomain)
        self.relayVerdict = try c.decode(RelayVerdict.self, forKey: .relayVerdict)
        self.lockedDown = try c.decodeIfPresent(Bool.self, forKey: .lockedDown) ?? false
        self.failingCertHash = try c.decodeIfPresent(String.self, forKey: .failingCertHash)
        self.coveringExceptionCertHash = try c.decodeIfPresent(String.self, forKey: .coveringExceptionCertHash)
        self.nonce = try c.decode(String.self, forKey: .nonce)
        self.issuedAt = try c.decode(Int64.self, forKey: .issuedAt)
    }
}

public enum BoxTrustStatus {
    public static let canonicalTag = "flagship/box-trust-status/v1"

    /// Freshness bound on a relayed report, mirroring `DaemonStatus`: older
    /// than this and the verdict is treated as stale (fail-open, no alarm).
    public static let maxReportAgeMs: Int64 = 7 * 24 * 60 * 60 * 1000

    public static func canonicalBytes(_ r: BoxTrustStatusReport) -> Data {
        let segments = [
            canonicalTag,
            r.serverDomain,
            r.relayVerdict.rawValue,
            r.lockedDown ? "1" : "0",
            r.failingCertHash ?? "",
            r.coveringExceptionCertHash ?? "",
            r.nonce,
            String(r.issuedAt)
        ]
        return Data(segments.joined(separator: "|").utf8)
    }

    /// Verify the box's signature under its (locally derived) STK pubkey.
    /// Returns false — never throws — on malformed input.
    public static func verify(_ r: BoxTrustStatusReport, signature: Data, stkPub: Data) -> Bool {
        guard signature.count == 64,
              let pub = try? Curve25519.Signing.PublicKey(rawRepresentation: stkPub)
        else { return false }
        return pub.isValidSignature(signature, for: canonicalBytes(r))
    }

    public static func verify(_ r: BoxTrustStatusReport, signatureHex: String, stkPub: Data) -> Bool {
        guard let sig = HexUtil.decode(signatureHex) else { return false }
        return verify(r, signature: sig, stkPub: stkPub)
    }
}

File: packages/protocol/src/boxTrustStatus.ts
/**
 * STK-signed box-trust-status report — the PER-BOX relay-trust verdict
 * primitive (maintainer-trust enforcement, "per-server" surfacing).
 *
 * Each box independently verifies the ServiceBlessing (relay-class trust)
 * it is handed by the `.services` hub. That verdict is genuinely per-box —
 * two boxes on the same account can disagree if one is handed a stale or
 * invalid blessing. The daemon emits this SIGNED report on its heartbeat;
 * `.com` relays the VERBATIM signed tuple + signature on /pods; a client
 * that derived the box STK locally re-verifies it end-to-end, so a rogue
 * `.com` can DROP the report but cannot FORGE one — the verdict a client
 * renders is the box's own word, not `.com`'s.
 *
 * This is a SIBLING of daemonStatus.ts — do NOT fold it into that report:
 * the daemon-status canonical bytes are pinned across three platforms and
 * carry cert-fingerprint pinning, an orthogonal concern with its own
 * cadence. Keep the two envelopes independent.
 *
 * Canonical bytes (one implementation, shared by the daemon heartbeat, the
 * control-plane relay, and every client; iOS/Android mirror this
 * byte-for-byte — see the pinned vector in tests/boxTrustStatus.test.ts):
 *
 *   flagship/box-trust-status/v1|<serverDomain>|<relayVerdict>|
 *   <lockedDown "1"|"0">|<failingCertHash or "">|
 *   <coveringExceptionCertHash or "">|<nonce>|<issuedAt>
 */
import { ed } from "./edSync.js";
import { legacyFieldGuard } from "./auth.js";
import { resolveMsgSigner, type MsgSigner } from "./canonicalBase.js";
import type { Bytes } from "./types.js";

/**
 * The box's verdict on the relay-class (`.services` hub) blessing it holds:
 *   trusted   — a currently-valid ServiceBlessing verified against the pin.
 *   untrusted — a blessing was expected but is missing/expired/invalid (RAISE).
 *   unknown   — trust enforcement is not configured / not yet evaluated
 *               (e.g. a transient network blip; stays fail-open, no alarm).
 */
export type RelayVerdict = "trusted" | "untrusted" | "unknown";

export interface BoxTrustStatusReport {
  serverDomain: string;
  relayVerdict: RelayVerdict;
  /** True when the box has entered relay-lockdown (data plane sealed). */
  lockedDown: boolean;
  /** relay-class cert-hash of the offending hub key, when untrusted. */
  failingCertHash: string | null;
  /** relay-class cert-hash of the owner TrustException that lifted the
   *  failing verdict, when an override is in force. */
  coveringExceptionCertHash: string | null;
  nonce: string;
  issuedAt: number;
}

const TAG_BOX_TRUST_STATUS = "flagship/box-trust-status/v1";

export function canonicalBoxTrustStatusReport(r: BoxTrustStatusReport): Bytes {
  legacyFieldGuard("serverDomain", r.serverDomain);
  legacyFieldGuard("relayVerdict", r.relayVerdict);
  if (r.failingCertHash !== null)
    legacyFieldGuard("failingCertHash", r.failingCertHash);
  if (r.coveringExceptionCertHash !== null)
    legacyFieldGuard("coveringExceptionCertHash", r.coveringExceptionCertHash);
  legacyFieldGuard("nonce", r.nonce);
  return new TextEncoder().encode(
    [
      TAG_BOX_TRUST_STATUS,
      r.serverDomain,
      r.relayVerdict,
      r.lockedDown ? "1" : "0",
      r.failingCertHash ?? "",
      r.coveringExceptionCertHash ?? "",
      r.nonce,
      String(r.issuedAt),
    ].join("|"),
  );
}

export function signBoxTrustStatusReport(
  r: BoxTrustStatusReport,
  identity: MsgSigner,
): Bytes {
  // Accepts the STK `Keypair` or a `sign(msg)` closure (custodian-backed);
  // signature bytes are identical either way.
  return resolveMsgSigner(identity)(canonicalBoxTrustStatusReport(r));
}

export function verifyBoxTrustStatusReport(
  r: BoxTrustStatusReport,
  sig: Bytes,
  stkPub: Bytes,
): boolean {
  try {
    return ed.verify(sig, canonicalBoxTrustStatusReport(r), stkPub);
  } catch {
    return false;
  }
}
